webgpu.h

```diff
@@ -0,0 +1,1025 @@
+    wgpuShaderModule module;
+    char const * entryPoint;
+} wgpuPipelineStageDescriptor;
+
+typedef struct wgpuRasterizationStateDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuFrontFace frontFace;
+    wgpuCullMode cullMode;
+    int32_t depthBias;
+    float depthBiasSlopeScale;
+    float depthBiasClamp;
+} wgpuRasterizationStateDescriptor;
+
+typedef struct wgpuRenderPassDepthStencilAttachmentDescriptor {
+    wgpuTextureView attachment;
+    wgpuLoadOp depthLoadOp;
+    wgpuStoreOp depthStoreOp;
+    float clearDepth;
+    wgpuLoadOp stencilLoadOp;
+    wgpuStoreOp stencilStoreOp;
+    uint32_t clearStencil;
+} wgpuRenderPassDepthStencilAttachmentDescriptor;
+
+typedef struct wgpuSamplerDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuAddressMode addressModeU;
+    wgpuAddressMode addressModeV;
+    wgpuAddressMode addressModeW;
+    wgpuFilterMode magFilter;
+    wgpuFilterMode minFilter;
+    wgpuFilterMode mipmapFilter;
+    float lodMinClamp;
+    float lodMaxClamp;
+    wgpuCompareFunction compare;
+} wgpuSamplerDescriptor;
+
+typedef struct wgpuShaderModuleDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    uint32_t codeSize;
+    uint32_t const * code;
+} wgpuShaderModuleDescriptor;
+
+typedef struct wgpuStencilStateFaceDescriptor {
+    wgpuCompareFunction compare;
+    wgpuStencilOperation failOp;
+    wgpuStencilOperation depthFailOp;
+    wgpuStencilOperation passOp;
+} wgpuStencilStateFaceDescriptor;
+
+typedef struct wgpuTextureViewDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuTextureFormat format;
+    wgpuTextureViewDimension dimension;
+    uint32_t baseMipLevel;
+    uint32_t mipLevelCount;
+    uint32_t baseArrayLayer;
+    uint32_t arrayLayerCount;
+    wgpuTextureAspect aspect;
+} wgpuTextureViewDescriptor;
+
+typedef struct wgpuVertexAttributeDescriptor {
+    uint32_t shaderLocation;
+    uint64_t offset;
+    wgpuVertexFormat format;
+} wgpuVertexAttributeDescriptor;
+
+typedef struct wgpuBindGroupDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuBindGroupLayout layout;
+    uint32_t bindingCount;
+    wgpuBindGroupBinding const * bindings;
+} wgpuBindGroupDescriptor;
+
+typedef struct wgpuBindGroupLayoutDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    uint32_t bindingCount;
+    wgpuBindGroupLayoutBinding const * bindings;
+} wgpuBindGroupLayoutDescriptor;
+
+typedef struct wgpuColorStateDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
+    wgpuTextureFormat format;
+    wgpuBlendDescriptor alphaBlend;
+    wgpuBlendDescriptor colorBlend;
+    wgpuColorWriteMask writeMask;
+} wgpuColorStateDescriptor;
+
+typedef struct wgpuComputePipelineDescriptor {
+    const void* nextInChain;
+    wgpuStructureType structureType;
```

Comment:
Do we really need these in the base descriptors? We could say that the structure types here are always what WebGPU version X.Y defines. We could require only those "nextInChain" things to provide the structure types.

Reply:
[1]

The `structureType` is useful in the base descriptor because it allows extensions to provide a completely alternative descriptor for some functions / parent descriptors. It also allows making WebGPU 2 descriptors by just making structureType be `WGPU_STRUCTURE_TYPE_WEBGPU2_DEFAULT`.

All your comments below about things being pointers are valid cosmetic concerns that happen when we require base descriptors to be swappable like this, but I think it is important in the long run. (like Vulkan forgot to do this for some descriptors and had to add them back in some VK_KHR_maintenance extensions, search for "2KHR" in vulkan.h)

Reply:
Ok, thanks for clarification.
Do *we* need to allow extensions to expect a completely different type? Do you have a good case in mind that would demonstrate why this is useful to us?

Reply:
I'm thinking about extensions that allow importing external resources, they could completely replace the descriptors with their own. Note that in C with designated initializers developers never need to specify the `WGPU_STRUCTURE_TYPE_DEFAULT` because it is 0 (and in C++ the structure type would have a default value).

Reply:
Ok, you want external resources. Can't use express them via the pNext chain? It doesn't look like it requires having the structure type in the base structs anyway.

Reply:
I got confirmation from Chris Forbes that there is no Vulkan extension that replaces the initial struct (so far).

I don't have a super good case in mind but am still very slightly leaning in favor of having `structureType` for initial structure.

Pros:

 - Structures are more uniform and tooling doesn't need to know the difference between initial structures and not initial.
 - We can replace the initial structure maybe one day for an extension or Webgpu 2
 - (more like the absence of a cons): no added developer friction thanks to C designated initializers, C++ and Rust defaults.

Cons:

 - Slightly larger types and one additional validation check.
 - Makes a couple structures pointed require one more indirection.

Reply:
Thanks for investigation!

I don't think there is a compelling evidence about the pros here:
  - tooling work isn't too complex - just a matter of an extra bit that you know beforehand
  - replacing an initial structure is not used in Vulkan, as you mentioned, so we'll unlikely need it either. I don't see how anything we'd want couldn't be done via `pNext` anyway?
  - Rust doesn't have defaults for *fields*.

Think about all the redundancy and clutter this would be adding to the header and docs... How about we come back to this question when it's time for V1 release?

Reply:
Sounds good. Note that this remove the need for `WGPUStructureType` at this point so I'll remove it.